### DISCUSSION TITLE
Support Enum as Security attribute/role

### DISCUSSION THREAD

**Discussion by sadikoff** [symfony/symfony#46006] [Q&A]
It can be nice to have the full support of Enums for security voters and maybe roles too. For now, in most places `$attribute` parameter is not type-hinted or `mixed` type is used depending on Symfony version, but
here https://github.com/symfony/symfony/blob/9cbc85389aaa9158943b582e56e4faf3d24b613b/src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php#L82
and here 
https://github.com/symfony/symfony/blob/9cbc85389aaa9158943b582e56e4faf3d24b613b/src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php#L88

`$attribute` should be string, and multiple voter implementations requires attribute to be `string` which not works if you pass Enum as attribute.

**Comment by stof** [comment]
Well, enums are meant to represent a *closed* list of values. In the security system, the list of permission attributes is unknown. It is an open list. So enums are not suited for that.

**Reply by sadikoff** [reply]
Yep, I got your thought, but what about inconsistency in parameters definition? I mean in some places `$attributes` are `mixed` and in some `string`?

**Reply by tarlepp** [reply]
imho - roles - is closed lists.

**Reply by DylanKas** [reply]
It's rarely the case when building application with a complex or dynamic role system

**Reply by joshpme** [reply]
With match expressions now supporting exhaustive match hinting in some IDEs.

And bits of code like this common in custom voters. 
```     
protected function supports($attribute, $subject) {
        $validChecks = [self::REVIEW, self::PAYMENT, self::REFUND, self::CANCEL, self::MODIFY, self::THANK];

        if(!in_array($attribute, $validChecks)) {
            return false;
        }
```

Attributes allowing the support of enum values. 

Supporting enums in voters is clear improvement.

**Reply by stof** [reply]
@tarlepp in your specific project, it likely is a closed list. But in `symfony/security-core`, it is not.
Each project using Symfony will want to have their *own* list of roles, not a closed list defined in `symfony/security-core`.

**Comment by Justinas-Jurciukonis** [comment]
I am using Roles as Enums, because it's finite list of values. But I find it very strange that up until `AccessDecisionManager::getVoters` I can pass anything as `$attributes`. E.g. `AuthorizationCheckerInterface->isGranted(AuthorizationEnum::ADMIN)`.

But when getting voter, it suddenly starts to work with only strings. At least accept [`Stringable`](https://www.php.net/manual/en/class.stringable.php) objects.

Now I have to do `->isGranted(AuthorizationEnum::ADMIN->value, AuthorizationEnum::ADMIN)` just for my voter to be able to work with Enums.

**Reply by stof** [reply]
Accepting stringable won't help you, as enums cannot be stringable.

**Comment by manuakasam** [comment]
I would like to get some clarification on this issue.

I would argue that most real life applications out there will have any form of constant holder for Roles / Permissions / similar tokens. When taking into considerations that database also holds values like this, then having them stored as an enum makes a hell of a lot of sense. 

I see absolutely no harm in supporting backedEnums for the isGranted functionality.

NOT doing so, honestly just promotes userland code errors. Why the hard stance against this?

**Reply by stof** [reply]
Enums provide a benefit to avoid errors when you can use them as the type of parameters to restrict things. But the AuthorizationCheckerInterface cannot use such application-specific enum as its parameter type, as it cannot depend on it.

PHP enums are representing a type with a *closed* list of values. But as far as `symfony/security-core` is concerned, permissions are an *open* list (which is what allows you to define your own permissions in your application) and so enums are not the right fit for that.

**Reply by Menelion** [reply]
@stof Christophe, sorry, but I still don't understand your point. I like the DRY principle (as many do, I bet). So in my app I have a set of permissions like that:

```php
enum ProjectPermission: string
{
    case View = 'view';
    case MarkAsDone = 'markAsDone';
}
```

To me it's a closed set of options, if I need more, I'll add, but it's pertinent to one entity and it's not open.  
So in my voter I did (it doesn't work, hence I googled for this discussion):

```php
    protected function supports(string $attribute, mixed $subject): bool
    {
        return in_array($attribute, ProjectPermission::cases(), true) && $subject instanceof Project;
    }
```

And, returning to the DRY principle, in my controller I could do:

```php
    #[Route('/{slug}', name: 'project_view', methods: ['GET'])]
    public function view(string $slug, Request $request, ProjectRepository $projectRepository): Response
    {
        $project = $projectRepository->findBySlug($slug);

        if (!$project || !$this->isGranted(ProjectPermission::View, $project)) {
            throw $this->createNotFoundException('This project does not exist');
        }

return $this->render('project/view.html.twig', [
            'project' => $project,
        ]);
    }
```

I think it's beautiful and clean. However, without enums I need to use constants, then do an array of all the constants for calling `in_array` in the `supports()` method, then use those constants in the controller… Okay, of course, but not as good, in my opinion.
I've been developing Symfony apps for more than five years but I'm quite new at hacking internal Symfony code. so, if there is something against Symfony philosophy here, I'd be glad to understand.
Thank you for reading!

**Comment by FactorSpeed** [comment]
What about using [BackedEnum::tryFrom](https://www.php.net/manual/fr/backedenum.tryfrom.php)

```php
#[IsGranted(attribute: Permission::SOMETHING->value)]
```

```php
protected function supports(string $attribute, mixed $subject): bool
{
    $user = $this->security->getUser();

    if (!$user instanceof UserInterface) {
        return false;
    }

    return Permission::tryFrom($attribute) instanceof Permission;
}
```

**Comment by rrajkomar** [comment]
I am by no means a reference is anything symfony-related but here are my two cents on this topic nonetheless.

I honestly don't see how allowing a `\StringBackedEnum|Expression|string` parameter instead of only `Expression|string` on the outer edge security-related attribute `IsGranted` could be an issue...

I.e, given a Role enum defined as the following : 
```
enum Role:string {
    ADMIN = 'ADMIN';
    USER = 'USER';
}
```
passing `Role::ADMIN` vs `'ADMIN'` to `\Symfony\Component\Security\Http\Attribute\IsGranted` is not that different...
It looks like the `\Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface::isGranted` already can support the enums without having to use the backing value anyway so the issue here is realy limited to the IsGranted attribute itself.

However I don't think forcing people to go through hoops by using an imposed string format with `Role::ADMIN->value` or (even worse looking) using the expression language with something like `enum(PATH\\TO\\NAMESPACE\\ROLE::ADMIN)` is a very smart path.

And forcing people to use constants instead of enums would definitely go against the exact argument used earlier : roles are supposed to be a closed list.
Therefore they should be supposed to be represented by enums

Also, simply allowing `\StringBackedEnum|Expression|string` as a type instead of `Expression|string`  does not cause any "closed list" issue (as previously mentioned) but allows for a definitely cleaner and safer code (no developer could inadvertendly use an invalid value since all valid values come from the project-defined enum).

All it would take is one line of code to automatically retrieve the value property if the passed element is a StringBackedEnum...
It would not even need to impact the core code as the conversion could happen on an outer edge element, in the IsGrantedAttributeListener class.

**Reply by stof** [reply]
`\StringBackedEnum` does not exist in PHP

**Reply by rrajkomar** [reply]
Indeed that is an interface only present in PHPStorm stubs but it doesn't change anything : 
the principle could still apply to a BackedEum (only int and string are possible as backing values) 
And nothing prevents use from having a role named ROLE_1 (allthough that wouldn't make much sense)

**Comment by crtl** [comment]
Is this implemented now?
I too cannot understand the arguments against it.
I cannot see how security-http would be coupled in any way by this because from my point of view you would just have something like:
```
if (is_backed_enum($attribute)) {
  $attribute = get_backed_enum_value($attribute);
}
```

in `IsGrantedAttributeListener` which would be 100% BC.

Coding principles shouldnt undermine the dev experience. 
I bet at least 99% of developer has some `Roles` class with a bunch of constants and their first thought when using symfony after php enums came out was: "Lets use enums for roles".

**Reply by mbabker** [reply]
> and their first thought when using symfony after php enums came out was: "Lets use enums for roles".

The problem here is that the enum is an application-level concern that the framework knows nothing about.  And supporting enums does nothing to improve the framework's internals in any meaningful way IMO.  In fact, after watching the PRs roll in to Laravel to support enums everywhere (its Auth component, the event dispatcher and event names, there's a draft PR for using them in cache keys, job queue names, resolving foreign key values in Eloquent models, connection and queue names in its message bus), it really makes me less eager to want to see enums supported in higher level framework components or packages.

We shouldn't be trying to force enums everywhere.  And for those places where an application wants to use an enum but their upstream code uses strings, those ugly extra few characters it takes to convert an enum to a string aren't hurting your application's type safety in any meaningful way.

**Reply by manuakasam** [reply]
THIS, honestly, is a reason I can get behind and it makes a lot of sense from this point of view. It is very easy for any developer, myself included in this case, to see how easy it would be to implement here. However, looking at the whole framework and enum'ifying all those strings and having $var=fromEnumOrString($var); everywhere, actually would be rather ugly.

**Reply by crtl** [reply]
I can totally understand the reasoning but this proposal was only about user roles because:
- IsGranted is declared `final` for some unknown reason (Ive already made a feature request for this https://github.com/symfony/symfony/issues/60896) which prevents reusing roles and its very common for `IsGranted` to be used multiple times with same args.
- Enums enable type hinting

In the end devs just dont want to write the same string every time throughout multiple parts of the app which is prone to errors.
Constants work but circumvents phps type hinting. Its just nicer to have a signature like `function addRole(UserRole $role)` instead of `function addRole(string $role)` which requires manual type checking/assertions.

Possible workaround for this would be if my feature request is accepted and `IsGranted` is made non-final again.

>We shouldn't be trying to force enums everywhere. And for those places where an application wants to use an enum but their upstream code uses strings, those ugly extra few characters it takes to convert an enum to a string aren't hurting your application's type safety in any meaningful way.

We dont want to force anything but allow it and typing literal strings is a source of bugs and circumvents auto completion.

> The problem here is that the enum is an application-level concern that the framework knows nothing about. And supporting enums does nothing to improve the framework's internals in any meaningful way IMO.

Those are implementation details which are none of the consumers concern. It may not improve the codebase but the usage of it because you dont have to type strings. I think the main issue here is enums missing `__toString`.

Currently the code base is implying that both is possible but in the end it isnt.

`VoterInterface` accepts `array $attributes` and its nowhere mentioned that it has to be `string[]`:
```php
    /**
     * Returns the vote for the given parameters.
     *
     * This method must return one of the following constants:
     * ACCESS_GRANTED, ACCESS_DENIED, or ACCESS_ABSTAIN.
     *
     * @param mixed $subject    The subject to secure
     * @param array $attributes An array of attributes associated with the method being invoked
     *
     * @return self::ACCESS_*
     */
    public function vote(TokenInterface $token, mixed $subject, array $attributes): int;
```

but when looking at `CacheableVoterInterface` which is extended by the base `Voter`, attribute has to be a string:
```php

namespace Symfony\Component\Security\Core\Authorization\Voter;

/**
 * Let voters expose the attributes and types they care about.
 *
 * By returning false to either `supportsAttribute` or `supportsType`, the
 * voter will never be called for the specified attribute or subject.
 *
 * @author Jérémy Derussé <jeremy@derusse.com>
 */
interface CacheableVoterInterface extends VoterInterface
{
    public function supportsAttribute(string $attribute): bool;

    /**
     * @param string $subjectType The type of the subject inferred by `get_class` or `get_debug_type`
     */
    public function supportsType(string $subjectType): bool;
}

```

Looking at this it seems the bottlenecks are 
- `CacheableVoterInterface` because it "wrongly" extends `VoterInterface` by limiting attributes to `string` instead of `mixed` like in `VoterInterface`
- `IsGranted` implementation limiting `$attribute` to `string|Expression`

**Reply by stof** [reply]
@crtl enums enable typing only when you target a specific enum. If the parameter type is `\BackedEnum` to support any backed enum, you loose all typing benefits.

> Looking at this it seems the issue is the `CacheableVoterInterface` is the problem because it "wrongly" extends `VoterInterface` by limiting attributes to `string` instead of `mixed` like in `VoterInterface`

Voting on non-string attributes is possible (the ExpressionVoter votes on an Expression value object). But only string attributes can be supported by a cacheable voter because of how the caching system works.

**Reply by crtl** [reply]
> enums enable typing only when you target a specific enum. If the parameter type is \BackedEnum to support any backed enum, you loose all typing benefits.

For clarification. I was referencing  typing in domain layer not the framework.

> Voting on non-string attributes is possible (the ExpressionVoter votes on an Expression value object). But only string attributes can be supported by a cacheable voter because of how the caching system works.

Thats true but it can be easily fixed by adding a new method to `CachableVoterInterface` to generate keys for attributes:

```php
public function getCacheKey(mixed $attribute): string;
```

Voter would only require simple changes to provide a default implementation.

**Reply by mbabker** [reply]
> > enums enable typing only when you target a specific enum. If the parameter type is \BackedEnum to support any backed enum, you loose all typing benefits.
> 
> For clarification. I was referencing typing in domain layer not the framework.

That's kind of the issue though.  For your domain to use your types consistently, the framework has to support your domain's type somehow.  The framework can't support an `App\Role` enum, it can support `BackedEnum` or `UnitEnum` (if you don't want to use backed enums).  But even if the framework fully supported enums in the authorization-related code, by the time you get down to your Voter implementation, you still can't rely on your domain types without type assertions or juggling.  So even if you could write `#[IsGranted(Role::Admin)` in your controller, you're still mapping framework types (strings, generic enums) to domain types (`App\Role` enum) in `Voter::supportsAttribute()` or `Voter::supports()`.

> * IsGranted is declared `final` for some unknown reason (Ive already made a feature request for this [[Security] Make IsGranted Attribute non final to allow usage of custom attributes #60896](https://github.com/symfony/symfony/issues/60896)) which prevents reusing roles and its very common for `IsGranted` to be used multiple times with same args.

It's always safer to start final and then open something up versus defaulting to a "everything has to be open" policy, especially with a platform as widely used as Symfony.  A fair number of attribute classes started final and were then opened through later PRs for very similar reasons.

> Voter would only require simple changes to provide a default implementation.

While this is true in theory, it would also force B/C breaking changes on everyone using the `CacheableVoterInterface` or the abstract `Voter` class because they'd all have to support more than string attributes.  So that is now a very legitimate concern to keep in mind, how does the impact of that change balance out against the convenience of allowing an application's enum types to flow through the authorization-related interfaces.

**Reply by stof** [reply]
Making voters generate the cache key does not work: it would require knowing which voter to ask for the cache key before being able to read the cache to know which voter supports that attribute, which is a chicken-and-egg problem.

**Reply by Amunak** [reply]
> That's kind of the issue though. For your domain to use your types consistently, the framework has to support your domain's type somehow.

No it doesn't; like, ideally it would, but having easy support for it in the application code is much more important.

I.e. having correct, strict typing for methods like `addRole` and whatnot is what's the most desirable here (as well as using it in place of strings in attributes and such); the framework having type-checking for them is very much secondary.

One option to make this way more versatile while also allowing implementation flexibility would be to completely ignore Enums, and instead allow a new Interface to be accepted by the annotation, voters, etc. For example, we could have a RoleInterface, that would have a single method returning string (called `asRoleName` or something similar), and then people could use whatever they want: an enum, a regular class, or literally anything.

**Reply by stof** [reply]
@Amunak you can already use whatever representation you want in your own code (strings, enum, a doctrine relation, etc...) for the roles, as long as you handle converting them to the string type expected by `symfony/security-core` when implementing `UserInterface::getRoles`

**Reply by Amunak** [reply]
@stof but that still doesn't allow you to also just use the app's implementation for the security attributes and such. This would be a nice way to unify it all, in my opinion.
